tests/suites/test_suite_pem.function

```diff
@@ -32,30 +33,82 @@ exit:
 }
 /* END_CASE */
 
+/* BEGIN_CASE depends_on:MBEDTLS_PEM_PARSE_C */
+void mbedtls_pem_strnstr( char * needle, int len, int res, int offset )
+{
+    static const char * haystack = "Lorem ipsum dolor sit amet";
+
+    if( len == -1 )
+        len = strlen( haystack );
+    char* str = mbedtls_pem_strnstr( haystack, needle, len );
+
+    if( res == 0 )
+    {
+        TEST_ASSERT( str == NULL );
+    }
+    else
+    {
+        TEST_ASSERT( str != NULL );
+        TEST_EQUAL( (size_t)offset, (size_t)(str - haystack) );
+    }
+}
+/* END_CASE */
+
 /* BEGIN_CASE depends_on:MBEDTLS_PEM_PARSE_C:MBEDTLS_AES_C:MBEDTLS_DES_C:MBEDTLS_MD5_C:MBEDTLS_CIPHER_MODE_CBC */
 void mbedtls_pem_read_buffer( char *header, char *footer, char *data,
                               char *pwd, int res, data_t *out )
 {
     mbedtls_pem_context ctx;
-    int ret;
+    const char *valid_data = "^^^\nTWJlZCBUTFM=\n$$$";
+    const char *valid_pwd = "Mbed TLS";
     size_t use_len = 0;
+    size_t data_len = strlen( data );
     size_t pwd_len = strlen( pwd );
     const unsigned char *buf;
+    unsigned char *copy = NULL;
+    size_t copy_length;
+    int copy_end = 0;
+    int ret;
 
     mbedtls_pem_init( &ctx );
 
     ret = mbedtls_pem_read_buffer( &ctx, header, footer, (unsigned char *)data,
                 (unsigned char *)pwd, pwd_len, &use_len );
     TEST_ASSERT( ret == res );
-    if( ret != 0 )
-        goto exit;
 
-    use_len = 0;
-    buf = mbedtls_pem_get_buffer( &ctx, &use_len );
-    TEST_EQUAL( use_len, out->len );
-    TEST_ASSERT( memcmp( out->x, buf, out->len ) == 0 );
+    if( ret == 0 )
+    {
+        use_len = 0;
+        buf = mbedtls_pem_get_buffer( &ctx, &use_len );
+        ASSERT_COMPARE( buf, use_len, out->x, out->len );
+    }
+
+    mbedtls_pem_free( &ctx );
+    ret = mbedtls_pem_read_buffer_with_len( &ctx, header, footer, (unsigned char *)data,
+                data_len, (unsigned char *)pwd, pwd_len, &use_len );
+    TEST_ASSERT( ret == res );
+
+    if( res != 0 )
+    {
+        // Here we make sure valid data beyond the nul byte does not make
+        // it work unexpectedly, because mbedtls_pem_read_buffer_with_len is
+        // not supposed to look beyond any nul byte contained in data.
+
+        copy_length = strlen(data) + 1 + strlen(valid_data);
+        ASSERT_ALLOC( copy, copy_length + 1 );
+
+        mbedtls_snprintf( (char *)copy, copy_length + 1, "%s%c%s%n",
+                    data, 0, (const char *)valid_data, &copy_end );
+        TEST_ASSERT( (size_t)copy_end == copy_length );
+
+        mbedtls_pem_free( &ctx );
+        ret = mbedtls_pem_read_buffer_with_len( &ctx, "^^^", "$$$", copy, (size_t)copy_end,
+                    (unsigned char *)valid_pwd, strlen( valid_pwd ), &use_len );
```

Comment:
Should also validate the returned value in use_len (i.e. that it's consistent with the previous calls to mbedtls_pem_read_buffer_with_len and mbedtls_pem_read_buffer).

Reply:
Not sure what to do here, it's not supposed to be consistent because we don't use the same header and footer, at least I think? This call always fails with MBEDTLS_ERR_PEM_NO_HEADER_FOOTER_PRESENT, so use_len is unchanged.

Reply:
Agreed, not sure what my original comment was suggesting.